### PR TITLE
Upgrade to CUDA 12

### PR DESCRIPTION
**Pull request by zhuohan123**
I can successfully build vllm with:

CUDA 12.2
Pytorch 2.1.0 (built with CUDA 12.1)
xformers 0.0.22.post7

Note that vllm cannot be successfully built on CUDA 12.1 because of [this issue](https://github.com/pybind/pybind11/issues/4606). This can be solved after pytorch upgrade its pybind dependecy.

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -43,14 +43,14 @@ jobs:
...
+
     strategy:
       fail-fast: false
       matrix:
           os: ['ubuntu-20.04']
           python-version: ['3.8', '3.9', '3.10', '3.11']
-          pytorch-version: ['2.0.1']
-          cuda-version: ['11.8'] # Github runner can't build anything older than 11.8
+          pytorch-version: ['2.1.0']
+          cuda-version: ['12.2'] # Github runner can't build anything older than 11.8
```
Comment (WoosukKwon) [inline .github/workflows/publish.yml:53]:
```suggestion
          cuda-version: ['12.2']
```
Could you check whether this works with CUDA 12.2?

**Comment by Tostino** [comment]
I will double check,  but pretty sure I built with 12.1 last week.

Will edit when I have a chance to look at my changes.

File: .github/workflows/publish.yml (hunk 1 above)
Comment (WoosukKwon) [reply .github/workflows/publish.yml:53]: @zhuohan123 Did you check this by any chance?

**Comment by WoosukKwon** [comment]
I think we have to check the following items before the release:
1. Does our GitHub workflow work for CUDA 12.2?
2. If you compile our package with CUDA 12.2, does it work when the user’s NVIDIA driver version is < 12.2 (e.g., like in colab)?
3. Bump up the NVIDIA PyTorch docker image version.

**Comment by casper-hansen** [comment]
Hi vLLM maintainers. I suggest maintaining compatibility with torch 2.0.1 and CUDA 11.8.0 for a few more versions. The way this would work is that you create two versions of the wheel:
- PyPi: torch 2.1.0 and CUDA 12.1.1 wheel
- GitHub release: additional torch 2.0.1 and CUDA 11.8.0 wheel

The idea is that people can still install vLLM from the GitHub release if they do not have the correct CUDA version yet. This can be achieved by the following:

```python
VLLM_VERSION = "0.2.2"
PYPI_BUILD = os.getenv("PYPI_BUILD", "0") == "1"

if not PYPI_BUILD:
    try:
        CUDA_VERSION = "".join(os.environ.get("CUDA_VERSION", torch.version.cuda).split("."))[:3]
        VLLM_VERSION += f"+cu{CUDA_VERSION}"
    except Exception as ex:
        raise RuntimeError("Your system must have an Nvidia GPU for installing AutoAWQ")
```

In the GitHub workflow, add a conditional on which and check if the current CUDA version being used to build is the same as the one you want to release on PyPi.

```
if ( $env:CUDA_VERSION -eq $env:PYPI_CUDA_VERSION ){
    $env:PYPI_BUILD = 1
}
```

For reference, you can look at `setup.py` and `build.yaml` in AutoAWQ.
https://github.com/casper-hansen/AutoAWQ/blob/main/.github/workflows/build.yaml

File: requirements.txt [hunk 2]
```diff
@@ -5,9 +5,9 @@ pandas  # Required for Ray data.
 pyarrow  # Required for Ray data.
 sentencepiece  # Required for LLaMA tokenizer.
 numpy
-torch == 2.0.1
+torch >= 2.1.0
 transformers >= 4.34.0  # Required for Mistral.
-xformers == 0.0.22  # Required for Mistral.
+xformers >= 0.0.22  # Required for Mistral.
```
Comment (WoosukKwon) [inline requirements.txt:10]:
```suggestion
xformers >= 0.0.22.post7  # Required for CUDA 12.1.
```

**Comment by WoosukKwon** [comment]
@zhuohan123 Let's merge this PR for the ease of development?

TODOs (in the next PR):
- [ ] Update Dockerfile
- [ ] Update the NVIDIA PyTorch docker image version in `installation.rst`
- [ ] Build vllm+cu118 for release (#1589)

**Review by WoosukKwon — APPROVED** [review]

**Pull request merged**